File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/ASTLocalVariableDeclarationTest.java
/**
 * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
 */

package net.sourceforge.pmd.lang.java.ast;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

public class ASTLocalVariableDeclarationTest extends BaseParserTest {

    @Test
    public void testSingleDimArray() {
        ASTLocalVariableDeclaration node = java.parse(TEST1).findDescendantsOfType(ASTLocalVariableDeclaration.class).get(0);
        assertEquals(1, node.getArrayDepth());
    }

    @Test
    public void testMultDimArray() {
        ASTLocalVariableDeclaration node = java.parse(TEST2).findDescendantsOfType(ASTLocalVariableDeclaration.class).get(0);
        assertEquals(2, node.getArrayDepth());
    }

    @Test
    public void testMultDimArraySplitBraces() {
        ASTLocalVariableDeclaration node = java.parse(TEST3).findDescendantsOfType(ASTLocalVariableDeclaration.class).get(0);
        assertEquals(3, node.getArrayDepth());
    }

    private static final String TEST1 = "class Foo {\n void bar() {int x[] = null;}\n}";

    private static final String TEST2 = "class Foo {\n void bar() {int x[][] = null;}\n}";

    private static final String TEST3 = "class Foo {\n void bar() {int[] x[][] = null;}\n}";
}

File: pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/ast/ASTLocalVariableDeclarationTest.kt
/*
 * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
 */


package net.sourceforge.pmd.lang.java.ast

import net.sourceforge.pmd.lang.ast.test.shouldBe
import net.sourceforge.pmd.lang.java.types.JPrimitiveType.PrimitiveTypeKind.INT

class ASTLocalVariableDeclarationTest : ParserTestSpec({

    parserTest("Extra dimensions") {

        inContext(StatementParsingCtx) {

            // int x[][] = null;
            // int[] x[][] = null;

            "int x@A[];" should parseAs {
                localVarDecl {

                    it::getModifiers shouldBe modifiers { }

                    primitiveType(INT)

                    varDeclarator {
                        variableId("x") {

                            it::isField shouldBe false

                            it::getExtraDimensions shouldBe child {
                                arrayDim {
                                    annotation("A")
                                }
                            }
                        }
                    }
                }
            }
        }
    }

})
